exonum/src/messages2/protocol.rs

```diff
@@ -0,0 +1,709 @@
+                                $(
+                                    $class::$type(ref t) => {
+                                        t.signed_message()
+                                    }
+                                ),+
+                            }
+                        }
+                    ),+
+                }
+            }
+        }
+    };
+}
+
+impl_protocol!{
+    /// Composition of every exonum protocol messages.
+    /// This messages used in network p2p communications.
+    SignedMessage => Protocol {
+        /// Exonum basic node messages.
+        0 => Service {
+            /// `RawTransaction` representation.
+            RawTransaction = 0,
+            /// Handshake to other node.
+            Connect = 1,
+            /// `Status` information of other node.
+            Status = 2,
+        },
+        /// Exonum consensus specific node messages.
+        1 => Consensus {
+            /// Consensus `Precommit` message.
+            Precommit = 0,
+            /// Consensus `Propose` message.
+            Propose = 1,
+            /// Consensus `Prevote` message.
+            Prevote = 2,
+        },
+        /// Exonum node responses.
+        2 => Responses {
+            /// Information about transactions, that sent as response to `TransactionsRequest`.
+            TransactionsResponse = 0,
+            /// Information about block, that sent as response to `BlockRequest`.
+            BlockResponse = 1,
+        },
+        /// Exonum node requests.
+        3 => Requests {
+
+            /// Request of some propose which hash is known.
+            ProposeRequest = 0,
+            /// Request of unknown transactions.
+            TransactionsRequest = 1,
+            /// Request of prevotes for some propose.
+            PrevotesRequest = 2,
+            /// Request of peer exchange.
+            PeersRequest = 3,
+            /// Request of some future block.
+            BlockRequest = 4,
+        },
+
+    }
+}
+
+impl Protocol {
+    /// Creates new protocol message.
+    ///
+    /// # Panics
+    ///
+    /// On serialization fail this method can panic.
+    pub fn new<T: ProtocolMessage>(
+        message: T,
+        author: PublicKey,
+        secret_key: &SecretKey,
+    ) -> Protocol {
+        T::into_protocol(Protocol::concrete(message, author, secret_key))
+    }
+
+    /// Creates new protocol message.
+    /// Return concrete `Message<T>`
+    ///
+    /// # Panics
+    ///
+    /// On serialization fail this method can panic.
+    pub fn concrete<T: ProtocolMessage>(
+        message: T,
+        author: PublicKey,
+        secret_key: &SecretKey,
+    ) -> Message<T> {
+        let value = message.serialize().expect("Couldn't serialize data.");
+        let (cls, typ) = T::message_type();
+        let signed = SignedMessage::new(cls, typ, value, author, secret_key);
+        T::into_message_from_parts(message, signed)
+    }
+    /// Checks buffer and return instance of `Protocol`.
```

Comment:
I guess we should also say that the buffer is parsed here.

Reply:
It's probably implementation specific. And user shouldn't care about this.  